File: app/src/main/java/org/mesonet/app/radar/RadarModel.java
package org.mesonet.app.radar;

import org.mesonet.app.caching.Cacher;
import org.mesonet.app.dependencyinjection.PerFragment;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;



public class RadarModel
{
    private Map<String, RadarDetailModel> mRadarDetailMap = new HashMap<>();



    public static class RadarDetailModel
    {
        private String state;
        private String name;
        private float latitude;
        private float longitude;
        private float latDelta;
        private float lonDelta;
        private float range;



        public float GetLatitude()
        {
            return latitude;
        }



        public float GetLongitude()
        {
            return longitude;
        }



        public float GetLatDelta()
        {
            return latDelta;
        }



        public float GetLonDelta()
        {
            return lonDelta;
        }



        public float GetRange()
        {
            return range;
        }



        public String GetName()
        {
            return name;
        }



        public static class Builder
        {
            private RadarDetailModel mResult = new RadarDetailModel();



            public void SetValue(String inKey, String inType, String inValue)
            {
                Field field = FindField(inKey);

                if(field != null)
                {
                    try {
                        field.setAccessible(true);
                        if(inType.equals("string") && field.getType() == String.class)
                            field.set(mResult, inValue);
                        else if(inType.equals("real") && field.getType() == float.class)
                            field.set(mResult, Float.parseFloat(inValue));
                    }
                    catch (IllegalAccessException e)
                    {
                        e.printStackTrace();
                    }

                    field.setAccessible(false);
                }
            }



            private Field FindField(String inName)
            {
                Field[] fields = RadarDetailModel.class.getDeclaredFields();

                if(fields != null)
                {
                    for (int i = 0; i < fields.length; i++)
                    {
                        if(fields[i].getName().equals(inName))
                            return fields[i];
                    }
                }

                return null;
            }



            public RadarDetailModel Build()
            {
                return mResult;
            }
        }
    }
}

File: app/src/main/java/org/mesonet/app/radar/RadarModel.kt
package org.mesonet.app.radar

import java.lang.reflect.Field
import java.util.HashMap


class RadarModel {
    private val mRadarDetailMap = HashMap<String, RadarDetailModel>()


    class RadarDetailModel {
        private val state: String? = null
        private val name: String? = null
        private val latitude: Float = 0.toFloat()
        private val longitude: Float = 0.toFloat()
        private val latDelta: Float = 0.toFloat()
        private val lonDelta: Float = 0.toFloat()
        private val range: Float = 0.toFloat()


        fun GetLatitude(): Float {
            return latitude
        }


        fun GetLongitude(): Float {
            return longitude
        }


        fun GetLatDelta(): Float {
            return latDelta
        }


        fun GetLonDelta(): Float {
            return lonDelta
        }


        fun GetRange(): Float {
            return range
        }


        fun GetName(): String? {
            return name
        }


        class Builder {
            private val mResult = RadarDetailModel()


            fun SetValue(inKey: String, inType: String, inValue: String) {
                val field = FindField(inKey)

                if (field != null) {
                    try {
                        field.isAccessible = true
                        if (inType == "string" && field.type == String::class.java)
                            field.set(mResult, inValue)
                        else if (inType == "real" && field.type == Float::class.javaPrimitiveType)
                            field.set(mResult, java.lang.Float.parseFloat(inValue))
                    } catch (e: IllegalAccessException) {
                        e.printStackTrace()
                    }

                    field.isAccessible = false
                }
            }


            private fun FindField(inName: String): Field? {
                val fields = RadarDetailModel::class.java.declaredFields

                if (fields != null) {
                    for (i in fields.indices) {
                        if (fields[i].name == inName)
                            return fields[i]
                    }
                }

                return null
            }


            fun Build(): RadarDetailModel {
                return mResult
            }
        }
    }
}
